### ISSUE TITLE
Inconsistent hostname verification applied when `filename` flag is used

### ISSUE THREAD

**Issue by atc0005** [atc0005/check-cert#315]
For the `lscert` and `check_cert` tools, using both the `filename` and `server` flags together is blocked by config validation. The `dns-name` flag is optional for both tools.

For `lscert`, the `VerifyHostname()` method call is performed *only* if the `server` flag is set and only if there are certs present in the chain:

https://github.com/atc0005/check-cert/blob/eb5eb94012842d984d49699cf0fee90f5fbddca3/cmd/lscert/main.go#L206-L226

If the optional `dns-name` flag is provided it is used for hostname verification, but the current implementation requires that the `server` flag is set; this implementation disables hostname verification if examining a certificate file. It is likely that this was an intentional decision; this behavior was already present in the initial prototype release, so I don't have a GH issue noting the intention behind the choice.

For `check_cert`, the logic to determine whether hostname verification is applied is more complex.

https://github.com/atc0005/check-cert/blob/eb5eb94012842d984d49699cf0fee90f5fbddca3/cmd/check_cert/main.go#L313-L321

1. We first assert that there is at least one certificate in the chain
2. We start with using the `server` flag value for later hostname verification (may be empty)
3. If the `dns-name` flag is used, we pick that value for hostname verification over the `server` flag value

https://github.com/atc0005/check-cert/blob/eb5eb94012842d984d49699cf0fee90f5fbddca3/cmd/check_cert/main.go#L331-L339

4. If the leaf cert has an empty SANs list and the user opted to skip hostname verification (when SANs list is empty) we emit a warning and then skip hostname verification

https://github.com/atc0005/check-cert/blob/eb5eb94012842d984d49699cf0fee90f5fbddca3/cmd/check_cert/main.go#L354

5. Regardless of whether the SANS list is populated, if the user didn't opt to skip hostname verification we do so here

https://github.com/atc0005/check-cert/blob/eb5eb94012842d984d49699cf0fee90f5fbddca3/cmd/check_cert/main.go#L367-L370

6. If there is a verification error we check if it was because the SANs list was empty and fail with verbose feedback to assist sysadmin with resolving the issue

https://github.com/atc0005/check-cert/blob/eb5eb94012842d984d49699cf0fee90f5fbddca3/cmd/check_cert/main.go#L404

7. If there was a verification error for another reason we fail with a recommendation to specify the correct site FQDN instead of a host FQDN

For `check_cert`, the first "gate" is whether a certificate is present in the chain (this is a little ironic as the block just above it fails the plugin if zero certs were found). Then, hostname verification is performed regardless of whether the `filename` or `server` flag was used. Instead of requiring that the `dns-name` flag is used if the `filename` flag was specified, hostname verification is performed anyway and *presumably* the user is guided to provide the `dns-name` flag based on the hostname verification error encountered.

**Comment by atc0005** [comment]
> For `lscert`, the `VerifyHostname()` method call is performed _only_ if the `server` flag is set and only if there are certs present in the chain

Question: *Should* `lscert` attempt to perform hostname verification for certificate files?

- If `dns-name` is specified, then the user is presumably looking to apply verification to the certificate.
- If it's not, then the user is likely just wanting to view the cert metadata and be told whether it is expired.

**Comment by atc0005** [comment]
For `check_cert`, we probably need to assert that `dns-name` is specified if `server` is not set *and* if we want to enforce hostname verification.

That said, it's probably a good idea to step back and decide which validations are optional, which are required and what triggers each of them.

All of this is surfaced with work picking back up on #2. See also #211.

**Comment by atc0005** [comment]
> > For `lscert`, the `VerifyHostname()` method call is performed _only_ if the `server` flag is set and only if there are certs present in the chain
> 
> Question: _Should_ `lscert` attempt to perform hostname verification for certificate files?
> 
> * If `dns-name` is specified, then the user is presumably looking to apply verification to the certificate.
> * If it's not, then the user is likely just wanting to view the cert metadata and be told whether it is expired.

This is worth going ahead and applying as a standalone PR.

Larger changes to cleanup logic in `check_cert` is likely best postponed until at least partial support for #211 is implemented.
